plenoptic/simulate/models/frontend.py

```diff
@@ -1,96 +1,414 @@
+    kernel_size:
+        Shape of convolutional kernel.
+    width_ratio_limit:
+        Sets a lower bound on the ratio of `surround_std` over `center_std`.
+        The surround Gaussian must be wider than the center Gaussian in order to be a
+        proper Difference of Gaussians. `surround_std` will be clamped to `ratio_limit`
+        times `center_std`.
+    amplitude_ratio:
+        Ratio of center/surround amplitude. Applied before filter normalization.
+    pad_mode:
+        Padding for convolution, defaults to "circular".
+    pretrained:
+        Whether or not to load model params estimated from [1]_. See Notes for details.
+    activation:
+        Activation function following linear and gain control operations.
+    apply_mask:
+        Whether or not to apply circular disk mask centered on the input image. This is
+        useful for synthesis methods like Eigendistortions to ensure that the
+        synthesized distortion will not appear in the periphery. See
+        `plenoptic.tools.signal.make_disk()` for details on how mask is created.
+
+    Notes
+    -----
+    These 12 parameters (standard deviations & scalar constants) were reverse-engineered
+    from model from [1]_, [2]_. Please use these pretrained weights at your own
+    discretion.
+
+    References
+    ----------
+    .. [1] A Berardino, J Ballé, V Laparra, EP Simoncelli, Eigen-distortions of
+        hierarchical representations, NeurIPS 2017; https://arxiv.org/abs/1710.02266
+    .. [2] http://www.cns.nyu.edu/~lcv/eigendistortions/ModelsIQA.html
+    """
+
+    def __init__(
+        self,
+        kernel_size: Union[int, Tuple[int, int]],
+        width_ratio_limit: float = 4.0,
+        amplitude_ratio: float = 1.25,
+        pad_mode: str = "circular",
+        pretrained=False,
+        activation: Callable[[Tensor], Tensor] = F.softplus,
+        apply_mask: bool = False,
+    ):
+        super().__init__()
+        kernel_size = (31, 31) if pretrained else kernel_size
+
+        self.on = LuminanceContrastGainControl(
+            kernel_size,
+            True,  # on_center = True
+            width_ratio_limit,
+            amplitude_ratio,
+            pad_mode,
+            activation
+        )
+
+        self.off = LuminanceContrastGainControl(
+            kernel_size,
+            False,  # on_center = False
+            width_ratio_limit,
+            amplitude_ratio,
+            pad_mode,
+            activation
+        )
+
+        if pretrained:
+            self.load_state_dict(self._pretrained_state_dict())
+
+        self.apply_mask = apply_mask
+        self._disk = None  # cached disk to apply to image
+
+    def forward(self, x: Tensor) -> Tensor:
+        y = torch.cat((self.on(x), self.off(x)), dim=1)
+
+        if self.apply_mask:
+            im_shape = x.shape[-2:]
+            if self._disk is None or self._disk.shape != im_shape:  # cache new mask
+                self._disk = make_disk(im_shape).to(x.device)
```

Comment:
I think there's an edge case where this will fail if they run the model once on the cpu, then move the image over to the gpu and run it again (or similarly run it on an image on the gpu, same size as the cpu image).

So should probably check device as well.

Reply:
Good catch. Fixed.